storage/remote/queue_manager.go

```diff
@@ -261,18 +301,23 @@ type QueueManager struct {
 	droppedSamplesTotal        prometheus.Counter
 	numShardsMetric            prometheus.Gauge
 	failedSamplesTotal         prometheus.Counter
+	failedMetadataTotal        prometheus.Counter
 	sentBatchDuration          prometheus.Observer
+	sentMetadataDuration       prometheus.Observer
 	succeededSamplesTotal      prometheus.Counter
+	succeededMetadataTotal     prometheus.Counter
 	retriedSamplesTotal        prometheus.Counter
+	retriedMetadataTotal       prometheus.Counter
 	shardCapacity              prometheus.Gauge
 	maxNumShards               prometheus.Gauge
 	minNumShards               prometheus.Gauge
 	desiredNumShards           prometheus.Gauge
 	bytesSent                  prometheus.Counter
+	metadataBytesSent          prometheus.Counter
 }
 
 // NewQueueManager builds a new QueueManager.
-func NewQueueManager(metrics *queueManagerMetrics, watcherMetrics *wal.WatcherMetrics, readerMetrics *wal.LiveReaderMetrics, logger log.Logger, walDir string, samplesIn *ewmaRate, cfg config.QueueConfig, externalLabels labels.Labels, relabelConfigs []*relabel.Config, client StorageClient, flushDeadline time.Duration) *QueueManager {
+func NewQueueManager(metrics *queueManagerMetrics, watcherMetrics *wal.WatcherMetrics, readerMetrics *wal.LiveReaderMetrics, logger log.Logger, walDir string, samplesIn *ewmaRate, mCfg config.MetadataConfig, cfg config.QueueConfig, externalLabels labels.Labels, relabelConfigs []*relabel.Config, client StorageClient, flushDeadline time.Duration, sm scrape.ReadyManager) *QueueManager {
```

Comment:
this 100% isn't your fault, but the number of parameters here makes me sad :slightly_frowning_face:

Reply:
I tried to fight this one, but it was no good. I just gave in. 😭

Reply:
oh, yeah, I don't think anything we could do here will make it upstream, unfortunately